akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -306,6 +306,69 @@ object GraphStages {
     override def toString: String = s"SingleSource($elem)"
   }
 
+  final class FutureFlattenSource[T, M](
+    val future: Future[Graph[SourceShape[T], M]])
+    extends GraphStageWithMaterializedValue[SourceShape[T], Future[M]] {
+
+    ReactiveStreamsCompliance.requireNonNullElement(future)
+
+    val out = Outlet[T]("futureFlatten.out")
+    val shape = SourceShape(out)
+
+    override def initialAttributes = DefaultAttributes.futureSource // TODO
+
+    def createLogicAndMaterializedValue(attr: Attributes): (GraphStageLogic, Future[M]) = {
+      val materialized = Promise[M]()
+
+      val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+        private val sinkIn = new SubSinkInlet[T]("FlattenMergeSink")
+
+        private val cb = getAsyncCallback[Try[Graph[SourceShape[T], M]]] {
+          case scala.util.Success(graph) ⇒ {
+            setHandler(out, this)
+            sinkIn.setHandler(this)
+
+            sinkIn.pull()
+
+            val src = Source.fromGraph(graph).watchTermination() { (m, done) ⇒
+              materialized.completeWith(
+                done.map(_ ⇒ m)(ExecutionContexts.sameThreadExecutionContext))
+            }
+
+            src.runWith(sinkIn.sink)(interpreter.subFusingMaterializer)
+          }
+
+          case scala.util.Failure(t) ⇒ failStage(t)
+        }.invoke _
+
+        setHandler(out, new OutHandler {
+          def onPull(): Unit =
+            future.onComplete(cb)(ExecutionContexts.sameThreadExecutionContext)
+        })
+
+        def onPush(): Unit = {
+          if (isAvailable(out)) {
+            push(out, sinkIn.grab())
+            sinkIn.pull()
+          }
+        }
+
+        def onPull(): Unit = {}
+
+        override def onUpstreamFinish(): Unit =
+          if (!sinkIn.isAvailable /* && isClosed(in) */ ) {
```

Comment:
Based on existing code, but I guess there it's useless. Wanted to add more tests to be sure.